### DISCUSSION TITLE
Request level decoder/encoder

### DISCUSSION THREAD

**Discussion by corysullivan** [kean/Get#61] [Ideas]
In some situations, a specific request requires a different decoder/encoder than the ones set in the underlying client. (particularly around dates😖). Would you consider a PR that adds an optional decoder/encoder override at the request level?
```swift
public struct Request<Response>: @unchecked Sendable {
    public var method: HTTPMethod
    public var url: URL?
    public var query: [(String, String?)]?
    public var body: Encodable?
    public var headers: [String: String]?
    public var decoder: JSONDecoder?
    public var encoder: JSONEncoder? 
    public var id: String?
}
```
Or can you suggest another way to work around this?

**Comment by kean** [comment]
Hi, @corysullivan. Yes, I agree that Get should provide a way to set a per-request decoder/encoder. However, I'm not sure what would be the best way to design the API that supports it. Adding it to `Request` will probably require adopting copy-on-write – too many fields that are either reference types or are backed by reference types (strings, dictionaries).

**Reply by kean** [reply]
[Nuke](https://github.com/kean/Nuke) solves this by providing a [`userInfo`](https://kean-docs.github.io/nuke/documentation/nuke/imagerequest) dictionary in `ImageRequest` and a myriad delegate methods to allow customizing the pipeline behavior where you can read `userInfo`.

**Reply by corysullivan** [reply]
Thanks, @kean.  I will look at the `ImageRequest` in Nuke. To be sure I understand, you are concerned with a performance hit from copying `Request` due to the inner reference types and consequently are hesitant to add more.

**Reply by kean** [reply]
Yes, you got it right. I also would like to keep the API small, hence the idea to use a delegate method for that. It could wok without modifying the `Request` type: you could provide a different decoder based on the response type or the request path.
